invisible_cities/cities/hypathia.py

```diff
@@ -0,0 +1,173 @@
+import tables as tb
+
+from functools import partial
+
+from .. database       import load_db
+
+from .. reco                  import sensor_functions     as sf
+from .. reco                  import tbl_functions        as tbl
+from .. reco                  import peak_functions       as pkf
+from .. core. random_sampling import NoiseSampler         as SiPMsNoiseSampler
+from .. core                  import system_of_units      as units
+from .. io  .        pmaps_io import          pmap_writer
+from .. io  .run_and_event_io import run_and_event_writer
+from .. io  .      trigger_io import       trigger_writer
+from .. io  . event_filter_io import  event_filter_writer
+
+from .. dataflow            import dataflow as fl
+from .. dataflow.dataflow   import push
+from .. dataflow.dataflow   import pipe
+from .. dataflow.dataflow   import sink
+
+from .  components import city
+from .  components import print_every
+from .  components import collect
+from .  components import copy_mc_info
+from .  components import zero_suppress_wfs
+from .  components import WfType
+from .  components import sensor_data
+from .  components import wf_from_files
+from .  components import check_empty_pmap
+from .  components import check_nonempty_indices
+from .  components import get_number_of_active_pmts
+from .  components import build_pmap
+from .  components import compute_and_write_pmaps
+from .  components import simulate_sipm_response
+from .  components import calibrate_sipms
+
+
+@city
+def hypathia(files_in, file_out, compression, event_range, print_mod, detector_db, run_number,
+             sipm_noise_cut, filter_padding, thr_sipm, thr_sipm_type, pmt_wfs_rebin, pmt_pe_rms,
+             s1_lmin, s1_lmax, s1_tmin, s1_tmax, s1_rebin_stride, s1_stride, thr_csum_s1,
+             s2_lmin, s2_lmax, s2_tmin, s2_tmax, s2_rebin_stride, s2_stride, thr_csum_s2, thr_sipm_s2,
+             pmt_samp_wid=25*units.ns, sipm_samp_wid=1*units.mus):
+    if   thr_sipm_type.lower() == "common":
+        # In this case, the threshold is a value in pes
+        sipm_thr = thr_sipm
+
+    elif thr_sipm_type.lower() == "individual":
+        # In this case, the threshold is a percentual value
+        noise_sampler = SiPMsNoiseSampler(detector_db, run_number)
+        sipm_thr      = noise_sampler.compute_thresholds(thr_sipm)
+
+    else:
+        raise ValueError(f"Unrecognized thr type: {thr_sipm_type}. "
+                          "Only valid options are 'common' and 'individual'")
+
+    #### Define data transformations
+    sd = sensor_data(files_in[0], WfType.mcrd)
+
+    # Raw WaveForm to Corrected WaveForm
+    mcrd_to_rwf      = fl.map(rebin_pmts(pmt_wfs_rebin),
+                              args = "pmt",
+                              out  = "rwf")
+
+    # Add single pe fluctuation to pmts
+    simulate_pmt = fl.map(partial(sf.charge_fluctuation, single_pe_rms=pmt_pe_rms),
+                          args = "rwf",
+                          out = "ccwfs")
```

Comment:
While the charge resolution is definitely the major thing to simulate here, I'm a bit worried that for the S1s that we maybe need a bit of electronic noise. Since they don't tend to be seen in all PMTs and the current search method sums all PMTs before looking there might be some possibility, at least at lower energies, to get fake ones or miss a real one. We don't read the Pedestal RMS from the mysql at the moment (and the values are a bit out of date anyway) but maybe it's worthwhile to at least simulate a nominal Gaussian noise? @jjgomezcadenas, @gonzaponte  ? The more complex electronics simulation can be left for when we hace more info and want to run diomira and irene instead of hypathia

Reply:
I'm afraid I can not discuss that, maybe @jjgomezcadenas or @gonzaponte can. In any case, it might be added in a later version of Hypathia.

Reply:
I ran a quick couple of tests on a couple of events from the simulations we've been using to test the new setup etc.

I just added a Gaussian noise with a sigma of ~1 ADC equivalent (1/20 pe) and another test with ~0.7 ADC equivalent and it does make a difference to the width and signal in the resultant S1s and S2s. The summed pmt signal in the S1 varies by up to ~30% and in the S2 by up to ~1.5%. Though I only ran a few events so it could be bad luck with the seeding.

It's maybe not hugely important but it's relatively easy to implement. We can maybe leave it for another PR though. I'll hold off on approving to see if there are any opinions.

Reply:
Simulating a Gaussian noise seems reasonable and easy to do. But if you prefer it can be done in a separate PR, maybe even reading the pedestal rms from the db,

Reply:
If we want to read from the database we'd need changes in load_db and to add the info to the DB. Since we don't have info on what the noise will be in NEXT100 other than that it's likely to be very similar to what we have in NEW. Maybe it's easier just to decide a representative number and put it in the hypathia configuration. The last time we explicitly measured it the mean over the 11 active PMTs at the time was 0.73 ADC. Maybe we could just add something along those lines.

That being said, the current status is more or less what @jmbenlloch used for his study + cosmetics. Maybe there's merit in accepting the PR as is and then getting someone to do a PR adding in a simple noise simulation. Thoughts?

Reply:
I would prefer to add the electronic noise in a different PR. Anyway, we should discuss (maybe through the email I sent a couple of weeks ago?) if we want to include it in the current studies of NEXT-100 or leave it for later.